cpp/src/metrics/pairwise_distance_correlation.cu

```diff
@@ -0,0 +1,76 @@
+
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <raft/distance/distance.cuh>
+#include <raft/handle.hpp>
+#include "pairwise_distance_correlation.cuh"
+
+namespace ML {
+
+namespace Metrics {
+void pairwise_distance_correlation(const raft::handle_t& handle,
+                                   const double* x,
+                                   const double* y,
+                                   double* dist,
+                                   int m,
+                                   int n,
+                                   int k,
+                                   raft::distance::DistanceType metric,
+                                   bool isRowMajor,
+                                   double metric_arg)
+{
+  // Allocate workspace
+  raft::mr::device::buffer<char> workspace(handle.get_device_allocator(), handle.get_stream(), 1);
+
+  // Call the distance function
+  switch (metric) {
+    case raft::distance::DistanceType::CorrelationExpanded:
+      raft::distance::
+        pairwise_distance_impl<double, int, raft::distance::DistanceType::CorrelationExpanded>(
+          x, y, dist, m, n, k, workspace, handle.get_stream(), isRowMajor);
+      break;
+    default: THROW("Unknown or unsupported distance metric '%d'!", (int)metric);
+  }
+}
+
+void pairwise_distance_correlation(const raft::handle_t& handle,
+                                   const float* x,
+                                   const float* y,
+                                   float* dist,
+                                   int m,
+                                   int n,
+                                   int k,
+                                   raft::distance::DistanceType metric,
+                                   bool isRowMajor,
+                                   float metric_arg)
+{
+  // Allocate workspace
+  raft::mr::device::buffer<char> workspace(handle.get_device_allocator(), handle.get_stream(), 1);
+
+  // Call the distance function
+  switch (metric) {
+    case raft::distance::DistanceType::CorrelationExpanded:
```

Comment:
These functions are named after the associated distance measure so I would call a function named `pairwise_distance_correlation` expecting it to compute the correlation distance. If it's not possible that this will ever compute a distance other than `correlation`, why are we passing the `DistanceType` into these functions at all? Rather than using an if condition, I would propose removing the `metric` argument altogether and just having the function compute the distance for which its name implies. What do you think?

Reply:
These functions are named after the associated distance measure so I would call a function named pairwise_distance_correlation expecting it to compute the correlation distance. If it's not possible that this will ever compute a distance other than correlation, why are we passing the DistanceType into these functions at all? Rather than using an if condition, I would propose removing the metric argument altogether and just having the function compute the distance for which its name implies. What do you think?

Reply:
Good point! I think the metric arg got propagated to all distances due to euclidean dist having multiple sub-implementations.
I've removed this redundant check & arg now for all non-euclidean metrics which doesn't have multiple implementations.